#50 enable export of timezone from metadata in to-mos

FWuellhorst opened the pull request:
This closes #50.

Todos:
- [x] Check correctnes
- [x] Add test
- [x] add epw output and possibly others

FWuellhorst committed:
#50 enable export of timezone from metadata in to-mos

FWuellhorst committed:
enable timezone export on epw and format functions to avoid local duplicates

FWuellhorst committed:
export timezone, use float type

FWuellhorst committed:
update requirements

FWuellhorst committed:
update requirements

FWuellhorst committed:
chore: temp

MartinRaetz reviewed (changes requested):
Thank you very much @FWuellhorst.

Can you confirm that the tests will cover sufficiently the new feature? Since its the new default the regression tests should be running on the new default.

FWuellhorst committed:
Merge branch 'development' into 50_timezone

# Conflicts:
#	aixweather/core_data_format_2_output_file/to_epw_energyplus.py
#	aixweather/core_data_format_2_output_file/to_mos_TMY3.py
#	aixweather/transformation_to_core_data/EPW.py

FWuellhorst committed:
rename comparison files and add shifted ones

FWuellhorst committed:
add timezone to meta_data

FWuellhorst committed:
rename to export_in_utc

FWuellhorst committed:
add info on summertime

FWuellhorst committed:
rm local return

FWuellhorst committed:
Add timezone to other imports and keep old test export-in-utc

FWuellhorst committed:
fix get_name

FWuellhorst committed:
fix name and timezone in tests

FWuellhorst committed:
export forecast in utc

FWuellhorst commented:
@MartinRaetz:  I checked locally in Dymola for a single test and the feature works well for Modelica.

@TobiasSpratte : Can you check the with E+ if the exported .epw with the timezone also works as expected, especially in combination with SPAWN and mos. I don't know how to access the weather results from E+ in Spawn.

Future development options: 
- [ ] Add a simple test model to at least check created mos files locally for different tilts and orientations, see screenshot
<img width="545" height="476" alt="image" src="https://github.com/user-attachments/assets/5123c3ff-dc8c-456e-b209-1ea55454d817" />

- [ ] Make all tests check both utc and export with timezone
- [ ] add an example with a different timezone (e.g. -6) to trigger edge cases for large shifting

What do you think @MartinRaetz ? I would do that in other issues to keep this one focused on the timezone aspect.

TobiasSpratte commented:
I have checked the exported .mos and .epw files in EnergyPlus, Spawn and Modelica and everything works as expected. Thank you @FWuellhorst !

FWuellhorst committed:
start dymola test work

FWuellhorst committed:
add workflow.yml

FWuellhorst committed:
use get_city_from_location, make city-name optional and mock tests to avoid HTTP errors in CI

FWuellhorst committed:
chore test rework with parameterized_classes

FWuellhorst committed:
change pytest capitalcase and rm redundant paper workflow

FWuellhorst committed:
fix marshmallow version

FWuellhorst committed:
rm python 3.13 from testing for now

FWuellhorst committed:
use newer mashmallow version

FWuellhorst committed:
increase version due to new function, add release workflow

MartinRaetz commented on aixweather/core_data_format_2_output_file/to_epw_energyplus.py:
```diff
@@ -460,6 +466,9 @@ def _format_data(df, start, stop, fillna):
     )
 
+    ### Shift to desired timezone
+    df = df.shift(periods=timezone, freq="h", axis=0)
```
Gehe davon aus, dass das klappt. Aber achte drauf, ob nach dem shift der Datensatz noch zum "richtigen" Zeitpunkt startet. Glaube für modelica und try wichtig, sonst nicht.

FWuellhorst replied:
Ich füge in diesem branch auch schon mos-simulationen hinzu, dann sieht man direkt den Effekt.

FWuellhorst committed:
trigger ci and remove blank lines for pep

MartinRaetz commented on aixweather/imports/DWD.py:
```diff
@@ -208,6 +208,7 @@ def import_meta_DWD_historical(station: str) -> utils_import.MetaData:
     meta.station_exists_since = station_data["BEGINN"]
     meta.station_exists_until = station_data["ENDE"]
     meta.input_source = "DWD Historical"
+    meta.timezone = 1  # Always in Germany
```
Bin mir hier nicht sicher. 1 ist ja nur winterzeit. Die Daten die man beim DWD lädt, waren (glaube ich) UTC. 
Erklär hier mal die Wirkung der Zeile, ansonsten müsste die Timezone auch in dem "dict" stehen. Ist aber alles bisschen was her.

MartinRaetz commented on aixweather/imports/ERC.py:
```diff
@@ -126,6 +126,7 @@ def import_meta_from_ERC() -> MetaData:
     meta.altitude = 230  # ca. 200m + building height 30m
     meta.latitude = 50.7893
     meta.longitude = 6.0516
+    meta.timezone = 1
```
hier gleiche frage. Oder ist die timezone nur für den export relevant, weil in meta?

MartinRaetz replied:
Ah, glaube das ist alles nur die standard spec für den Export, richtig? Wenn das so ist passt!

FWuellhorst replied:
Ja genau, nur bei epw wird das auch für den import genutzt. Habe ich im Kommentare hinzugefügt.

FWuellhorst committed:
rm gitlab-ci.yml

MartinRaetz reviewed (approved):
Thank you for this major PR! I think its a very helpful modification to meet the users expectations!

FWuellhorst committed:
Merge pull request #81 from RWTH-EBC/80_actions [PYPI-RELEASE]

add workflow.yml

FWuellhorst committed:
Merge remote-tracking branch 'origin/main' into 50_timezone

# Conflicts:
#	aixweather/imports/TRY.py
#	tests/test_TRY.py

FWuellhorst committed:
document timezone usage

FWuellhorst committed:
add new reference results

FWuellhorst committed:
rm time-sleep

FWuellhorst committed:
add new dymola tests with reference results

FWuellhorst committed:
fix name

FWuellhorst committed:
fix skip

FWuellhorst committed:
add summer day test as well

FWuellhorst committed:
delete files again

FWuellhorst committed:
use critical warning instead of error

FWuellhorst committed:
fix comparison

FWuellhorst committed:
fix comparison

FWuellhorst committed:
fix comparison

FWuellhorst committed:
better error logs

FWuellhorst committed:
test if the regression test actually works

FWuellhorst committed:
fix error msg

FWuellhorst committed:
Raise warnings in case timezone differs to much. revert manual change as test works

FWuellhorst committed:
add defaults to regression results

FWuellhorst commented:
@MartinRaetz @TobiasSpratte 
I added regression tests which simulate the generated time-shifted mos in Modelica on different tilts and directions. You can create these and more plots with the new test_mos_simulation, executed as a plain python script locally. 

## Tilted roof first day of year diffuse radiation
<img width="640" height="480" alt="plots_summer=False_HDifTilRoo" src="https://github.com/user-attachments/assets/06b22b82-49d2-4598-a6a5-83180cbb53b1" />

## Tilted roof in summer direct radiation
<img width="640" height="480" alt="plots_summer=True_HDirTilRoo" src="https://github.com/user-attachments/assets/bd605806-3f98-474c-9394-1363af065e25" />


Overall, when exporting time zones shifts are valid. Shifting larger than 6-8 hours compared to the imported data, the values become higher in the plot, which is caused by the values of the next day in the original data. 
Still, I stored the _imported_timezone (which is set in the import classes with "set_imported_timezone" and added a warning if someone changes the default timezone of the import.

FWuellhorst committed:
update warning
